sdk/include/opentelemetry/sdk/logs/read_write_log_record.h

```diff
@@ -0,0 +1,190 @@
+   */
+  void SetBody(const opentelemetry::common::AttributeValue &message) noexcept override;
+
+  /**
+   * Get body field of this log.
+   * @return the body field for this log.
+   */
+  const opentelemetry::common::AttributeValue &GetBody() const noexcept override;
+
+  /**
+   * Set the trace id for this log.
+   * @param trace_id the trace id to set
+   */
+  void SetTraceId(opentelemetry::trace::TraceId trace_id) noexcept override;
+
+  /**
+   * Get the trace id of this log.
+   * @return the trace id of this log
+   */
+  const opentelemetry::trace::TraceId &GetTraceId() const noexcept override;
+
+  /**
+   * Set the span id for this log.
+   * @param span_id the span id to set
+   */
+  void SetSpanId(opentelemetry::trace::SpanId span_id) noexcept override;
+
+  /**
+   * Get the span id of this log.
+   * @return the span id of this log
+   */
+  const opentelemetry::trace::SpanId &GetSpanId() const noexcept override;
+
+  /**
+   * Inject trace_flags for this log.
+   * @param trace_flags the trace flags to set
+   */
+  void SetTraceFlags(opentelemetry::trace::TraceFlags trace_flags) noexcept override;
+
+  /**
+   * Inject trace_flags of this log.
+   * @return trace_flags of this log
+   */
+  const opentelemetry::trace::TraceFlags &GetTraceFlags() const noexcept override;
+
+  /**
+   * Set an attribute of a log.
+   * @param key the name of the attribute
+   * @param value the attribute value
+   */
+  void SetAttribute(nostd::string_view key,
+                    const opentelemetry::common::AttributeValue &value) noexcept override;
+
+  /**
+   * Get attributes of this log.
+   * @return the body field of this log
+   */
+  const std::unordered_map<std::string, opentelemetry::common::AttributeValue> &GetAttributes()
+      const noexcept override;
+
+  /**
+   * Get resource of this log
+   * @return the resource of this log
+   */
+  const opentelemetry::sdk::resource::Resource &GetResource() const noexcept override;
+
+  /**
+   * Set Resource of this log
+   * @param Resource the resource to set
+   */
+  void SetResource(const opentelemetry::sdk::resource::Resource &resource) noexcept;
+
+  /**
+   * Get instrumentation_scope of this log.
+   * @return  the instrumentation_scope of this log
+   */
+  const opentelemetry::sdk::instrumentationscope::InstrumentationScope &GetInstrumentationScope()
+      const noexcept override;
+
+  /**
+   * Set instrumentation_scope for this log.
+   * @param instrumentation_scope the instrumentation scope to set
+   */
+  void SetInstrumentationScope(const opentelemetry::sdk::instrumentationscope::InstrumentationScope
+                                   &instrumentation_scope) noexcept;
+
+private:
+  // Default values are set by the respective data structures' constructors for all fields,
+  // except the severity field, which must be set manually (an enum with no default value).
+  opentelemetry::logs::Severity severity_                 = opentelemetry::logs::Severity::kInvalid;
```

Comment:
So, these LogRecord fields (severity, body, attributes) would be copied/stored in the SDK before exporting them, unlike earlier when they were directly serialized in the exporter's format ?

Reply:
Yes, according to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/api.md#logrecord , [LogRecords](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/api.md#emit-logrecord) MUST be able to set these following fields.
So I think we can only store them into ReadWriteLogRecord, and we can just store `string_view` and integers here to reduce copy.

Reply:
Theonly advantage I see with having ReadableLogRecord, and ReadWriteLogRecord is that we can now easily define the getter methods for logRecord as required by specs. But I am not seeing the use-case of these getter methods. For traces too, we deviated from specs by not have read/write spans and directly uses recordable to serialize the data as a tradeoff for performance.

Reply:
`ReadableLogRecord` and `ReadWriteLogRecord` is defined in SDK, while `LogRecord` is defined in API. They are just like `Span` in API and SDK, and `logs::Recordable` is just like the `trace::Recordable`.
I believe which cost more is store data into `proto::v1::LogRecord` in OTLP and convert data into json in ETW and ES exporters.To reduce the cost, I try to only keep string_view and AttributeValue in LogRecord classes, these are all just a reference to the real data without lifetime(There is a except, attributes are store in a `unordered_map<std::string, AttributeValue>` now, maybe we can use a reference to `KeyValueIterable` instead of using `unordered_map`, but it may make it difficult to set attribute one by one and add `event.name` when emit a record with event), The getter methods are used to pick datas from LogRecord and set them into `Recordable` now.

Reply:
I think it easier to add or remove fields if log content changes in the future with `LogRecord` in API.

Reply:
Storing non-owning objects can have undesirable issues if these objects get out of scope before they get serialized by the exporter. I am still not seeing the benefit of storing them for implementing getter methods unless there is a use case where these getters would be used as public API.

Reply:
The exporter should create it's own `Recordable` object from `ReadableLogRecord` or `ReadWriteLogRecord` and the `Recordable` of exporter should own objects, just like the old implementation.The mainly different between old and new implementation is the old implementation set data fields one by one and the new implementation collect all data into `ReadWriteLogRecord` the lifetime and data copy times will not change, just like we talked before in #1691 .
`ReadableLogRecord` is not necessary but I think it's better to follow [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.15.0/specification/logs/sdk.md#readablelogrecord). Do you mean we can delete `ReadWriteLogRecord` and just keep `ReadWriteLogRecord` ?

Reply:
> does it means these getters would be able to be used as public API?

Yes, these getters should be available as public API.